library/pk.c

```diff
@@ -579,6 +577,284 @@ int mbedtls_pk_get_psa_attributes(const mbedtls_pk_context *pk,
+            ret = PSA_PK_TO_MBEDTLS_ERR(psa_import_key(attributes,
+                                                       key_data, key_length,
+                                                       key_id));
+            mbedtls_platform_zeroize(key_data, key_length);
+            return ret;
+        }
+#endif /* MBEDTLS_RSA_C */
+
+#if defined(MBEDTLS_PK_HAVE_ECC_KEYS)
+        case MBEDTLS_PK_ECKEY:
+        case MBEDTLS_PK_ECKEY_DH:
+        case MBEDTLS_PK_ECDSA:
+        {
+            /* We need to check the curve family, otherwise the import could
+             * succeed with nonsensical data.
+             * We don't check the bit-size: it's optional in attributes,
+             * and if it's specified, psa_import_key() will know from the key
+             * data length and will check that the bit-size matches. */
+            psa_key_type_t to_type = psa_get_key_type(attributes);
+#if defined(MBEDTLS_PK_USE_PSA_EC_DATA)
+            psa_ecc_family_t from_family = pk->ec_family;
+#else /* MBEDTLS_PK_USE_PSA_EC_DATA */
+            /* We're only reading the key, but mbedtls_ecp_write_key()
+             * is missing a const annotation on its key parameter, so
+             * we need the non-const accessor here. */
+            mbedtls_ecp_keypair *ec = mbedtls_pk_ec_rw(*pk);
+            size_t from_bits = 0;
+            psa_ecc_family_t from_family = mbedtls_ecc_group_to_psa(ec->grp.id,
+                                                                    &from_bits);
+#endif /* MBEDTLS_PK_USE_PSA_EC_DATA */
+            if (to_type != PSA_KEY_TYPE_ECC_KEY_PAIR(from_family)) {
+                return MBEDTLS_ERR_PK_TYPE_MISMATCH;
+            }
+
+#if defined(MBEDTLS_PK_USE_PSA_EC_DATA)
+            if (mbedtls_svc_key_id_is_null(pk->priv_id)) {
+                /* We have a public key and want a key pair. */
+                return MBEDTLS_ERR_PK_TYPE_MISMATCH;
+            }
+            return copy_into_psa(pk->priv_id, attributes, key_id);
+#else /* MBEDTLS_PK_USE_PSA_EC_DATA */
+            if (ec->d.n == 0) {
+                /* Private key not set. Assume the input is a public key only.
+                 * (The other possibility is that it's an incomplete object
+                 * where the group is set but neither the public key nor
+                 * the private key. This is not possible through ecp.h
+                 * functions, so we don't bother reporting a more suitable
+                 * error in that case.) */
+                return MBEDTLS_ERR_PK_TYPE_MISMATCH;
+            }
+            unsigned char key_buffer[PSA_BITS_TO_BYTES(PSA_VENDOR_ECC_MAX_CURVE_BITS)];
+            int ret = mbedtls_ecp_write_key(ec,
+                                            key_buffer, sizeof(key_buffer));
+            if (ret < 0) {
+                return ret;
+            }
+            size_t key_length = PSA_BITS_TO_BYTES(ec->grp.nbits);
+            unsigned char *key_data = key_buffer + sizeof(key_buffer) - key_length;
```

Comment:
Hmm, I found this surprising. I would have been less surprised by
```
size_t key_length = PSA_BITS_TO_BYTES(ec->grp.nbits);
int ret = mbedtls_ecp_write_key(ec, key_buffer, key_length);
// handle error
// call psa_import_key using key_buffer and key_length
```
though arguably it's less safe.

Actually I'm not sure the code as it stands is correct for Montgomery curves: here the private key is written little-endian so it's going to be zero-padded on the _right_ so the line I'm commenting on should not work. Unless I'm more confused than I think I am.

(I'm now thinking the behaviour of `mbedtls_ecp_write_key()` is silly. Not convenient to change in 3.x as it's a public function, but we might want to make it more sensible in 4.x.)

Reply:
Hmm, I think you're right. `mbedtls_ecp_write_key` is really hard to use, and this is not documented at all!

This brings up a defect in the testing: `test_suite_pk` focuses on metadata, and in `test_suite_pkparse` there's a check of internal consistency of the resulting key (`mbedtls_pk_import_into_psa` is followed by `mbedtls_test_psa_exercise_key`), but there is no testing that the key material in the PSA key is what we expect. I'll think about how to test that.

(I've pushed an update that should address all your comments except this one.)

Reply:
> I'll think about how to test that.

It's awkward due to the different representations, but we can export/write both keys in the same format and check that they match.

Also, we have two pk objects, so we can sign/encrypt with one and verify/decrypt with the other, or do a key agreement between the two. The situation is the same with `mbedtls_pk_copy_from_psa`, and I see that Valerio put this testing in #8774.

I'd prefer to [add an ECP function with a new interface](https://github.com/Mbed-TLS/mbedtls/issues/8835). But I don't want to expand the scope of this pull request any further, so I'll use your workaround.

The pair consistency testing code can be the same for `mbedtls_pk_copy_from_psa` and `mbedtls_pk_import_into_psa`, so I'd like to merge Valerio's test code, and then make a follow-up that uses this consistency testing for the other direction as well.

Reply:
Thinking further, all we want to know is that the two key objects (one in PK, one in PSA) have the same key material. Or, if one is a key pair and one is a public key, we want to know that the public parts are the same. We have other ways of testing that each object is internally consistent. So it's enough to export and compare the two public keys. The nice part is that this works regardless of the policy on the PSA key.

Reply:
I added a public key comparison test, and it caught the bug you noticed with Montgomery keys.